File: leetcode/js/count-substrings-that-satisfy-k-constraint-i.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var countKConstraintSubstrings = function (s, k) {
  let result = 0;

  for (let i = 1; i <= s.length; i++) {
    for (let j = 0; j <= s.length - i; j++) {
      const sub = s.slice(j, j + i);

      let count0 = 0;
      let count1 = 0;

      for (const c of sub) {
        if (c === "0") count0++;
        if (c === "1") count1++;
      }

      if (count0 <= k || count1 <= k) {
        result++;
      }
    }
  }

  return result;
};

File: leetcode/rust/count-substrings-that-satisfy-k-constraint-i.rs
impl Solution {
    pub fn count_k_constraint_substrings(s: String, k: i32) -> i32 {
        let mut left = 0;
        let mut count0 = 0;
        let mut count1 = 0;
        let mut result = 0;
        let s_chars: Vec<char> = s.chars().collect();

        for right in 0..s_chars.len() {
            match s_chars[right] {
                '0' => count0 += 1,
                '1' => count1 += 1,
                _ => (),
            };

            while count0 > k && count1 > k {
                match s_chars[left] {
                    '0' => count0 -= 1,
                    '1' => count1 -= 1,
                    _ => (),
                };
                left += 1;
            }

            result += right - left + 1;
        }

        result as i32
    }
}
